Fix mirror fetcher threads

fvaleri opened the pull request:
This patch ensures exactly one mirror fetcher thread per source partition, even during leader elections in either cluster.

Key fixes:

  ~~1. Duplicate fetchers: `applyLocalLeadersDelta` stops mirror fetchers to avoid having two fetcher threads for the same partition in two different destination brokers after a leader election.~~
  2. Stale bootstrap fetchers: `getRemotePartitionLeader` triggers synchronous metadata refresh when cache is empty, preventing invalid fetchers from persisting.
  3. Source election handling: Introduced `MirrorFetcherThread` to route add and remove operations to mirrorFetcherManager, enabling fetcher recreation when source leaders change.
  4. Negative epoch fix: Remote leaders skip `makeFollower` and manually track epoch changes to avoid epoch reset (addresses negative epoch issue).

Apologies for the refactoring noise (mirror package, class renames), but it improves separation of concerns, plus brings in some good comments.

fvaleri committed:
Fix remote leader epoch initialization

ReplicaManager.applyLocalFollowersDelta skips calling partition.makeFollower for partitions
with remote leader because we want to make consumer-like fetches. The problem is that makeFollower
also sets the partition's leader epoch, so it remains unitialized at -1.

With negative epoch, UnifiedLog.analyzeAndValidateRecords skips the batches in a tigh loop
causing high CPU consumption and logs flooded with:

[2025-09-24 14:00:51,999] INFO [UnifiedLog partition=my-topic-0, dir=/home/fvaleri/Documents/kafka/build/test/server2/data] Skipping batch RecordBatch(magic=2, offsets=[5, 5], sequence=[1, 1], partitionLeaderEpoch=0, isTransactional=false, isControlBatch=false, compression=none, timestampType=CreateTime, crc=1079926212) from an origin of REPLICATION because its partition leader epoch 0 is higher than the replica's current leader epoch -1 (org.apache.kafka.storage.internals.log.UnifiedLog)

It can be reproduced having multiple brokers and triggering leader elections.

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented:
@fvaleri , thanks for identifying this issue! I hope the intra-cluster replication can work as usual, no need to go through the consumer-fetch way. Do you think it is feasible? That is, only the leader in the target cluster go with the consumer-like fetch to the source cluster leader. All other followers should go with the normal logic, including `makeFollower`. I need some time to think about it, but want to hear your thought.

fvaleri committed:
Cleanups

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

showuon commented:
@fvaleri , is this PR ready for review?

fvaleri commented:
No, still playing with it. Will try to post an update later today.

fvaleri committed:
Fix mirror fetcher threads

This commit ensures exactly one mirror fetcher thread per source partition, even during leader elections in either cluster.

Key fixes:

1. Duplicate fetchers: `applyLocalLeadersDelta` stops mirror fetchers to avoid having two fetcher threads for the same partition in two different destination brokers after a leader election.
2. Stale bootstrap fetchers: `getRemotePartitionLeader` triggers synchronous metadata refresh when cache is empty, preventing invalid fetchers from persisting.
3. Source election handling: Introduced `MirrorFetcherThread` to route add and remove operations to mirrorFetcherManager, enabling fetcher recreation when source leaders change.
4. Negative epoch fix: Remote leaders skip makeFollower and manually track epoch changes to avoid epoch reset (addresses negative epoch issue).

Apologies for the refactoring noise (mirror package, class renames), but it improves separation of concerns, plus brings in some good comments.

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri commented:
@showuon you were right, there are other issues regarding remote fetching. PR description updated. Let me know what you think. Thanks.

showuon commented:
@fvaleri , thanks for the PR. But I have to let you know that you are not trying hard to let the reviewers review your PR easily. In this PR, you tried to fix 4 issues and mixed in one PR together. For a reviewer, how does he/she know where is the fix for issue 1, where is for issue 2,.... Furthermore, you also know you includes the refactoring noise, which is another problem make the PR hard to get reviewed. 

Comments:
1. Split this PR into small PRs to separate for different issues.
2. The PR is originally to fix issue 4, and we had a discussion about the possible problem for the current fix, do you still remember? 

Let me go through them again here, and let you know why I think this is an important issue and we need to treat it carefully, instead of mix other fixes/refactoring together.

a. The original problem is that when the target cluster follower fetch from the local leader, it'll get the exception:
```
Skipping batch RecordBatch... from an origin of REPLICATION because its partition leader epoch 0 is higher than the replica's current leader epoch -1 (org.apache.kafka.storage.internals.log.UnifiedLog)
```
b. The problem is because the leader epoch in the target cluster (-1) is smaller than the one in the batch (0).
c. We can update the leader epoch in the target cluster each time when there is a leadership change in the target cluster, that will fix the (-1) issue for sure. But this issue is not only about the -1 issue, it's about the nature of leader epoch is not identical between the source cluster and the target cluster.
d. Current solution will make the leader epoch in the target cluster not be (-1) anymore. So suppose it's (0) now. But when the batch fetched from the source cluster contains leader epoch = (1), the exception still be thrown.
e. So based on that, we need a better solution to fix this problem, in one PR.

Thank you again for your effort.

fvaleri commented:
> Split this PR into small PRs to separate for different issues.

This PR is the result of me exploring current code and finding a bunch of issues. I will open dedicated PR for refactoring and then one PR for each issue.

> d. Current solution will make the leader epoch in the target cluster not be (-1) anymore. So suppose it's (0) now. But when the batch fetched from the source cluster contains leader epoch = (1), the exception still be thrown.

Right. I guess I didn't fully get what you meant in our meeting and previous comment. Will think more about this and create a dedicated PR.

fvaleri commented:
Closing this one in favor of dedicated PRs.
